src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs

```diff
@@ -75,16 +93,68 @@ void tryCompileDeepFluentCalls(int depth)
 }");
 
                 var source = builder.ToString();
-
-                var thread = new System.Threading.Thread(() =>
+                RunInThread(() =>
                 {
                     var options = new CSharpCompilationOptions(outputKind: OutputKind.DynamicallyLinkedLibrary, concurrentBuild: false);
                     var compilation = CreateCompilation(source, options: options);
                     compilation.VerifyDiagnostics();
                     compilation.EmitToArray();
-                }, 0);
-                thread.Start();
-                thread.Join();
+                });
+            }
+        }
+
+        [Fact]
+        [WorkItem(33909, "https://github.com/dotnet/roslyn/issues/33909")]
+        public void DeeplyNestedGeneric()
+        {
+            int nestingLevel = 500;
+            var builder = new StringBuilder();
+            builder.AppendLine(@"
+using System;
+
+public class Test
+{
+    public static int Main()
+    {
+        #pragma warning disable 219
+");
+
+            for (var i = 0; i < nestingLevel; i++)
+            {
+                if (i > 0)
+                {
+                    builder.Append('.');
+                }
+                builder.Append($"MyStruct{i}<int>");
+            }
+
+            builder.AppendLine(" local;");
+            builder.AppendLine(@"
+        #pragma warning restore 219
+        Console.WriteLine(""Pass"");
+    }
+}");
+
+            genType(0);
+
+            var source = builder.ToString();
+
+            RunInThread(() =>
+            {
+                var compilation = CreateCompilation(source, options: TestOptions.DebugExe);
+                CompileAndVerify(compilation, expectedOutput: "Pass");
+            });
+
+            void genType(int level)
+            {
+                if (level >= nestingLevel)
+                {
+                    return;
+                }
+
+                builder.AppendLine($"public struct MyStruct{level}<T{level}> {{");
+                genType(level + 1);
```

Comment:
>genType(level + 1); [](start = 16, length = 19)

Are we likely to exceed the stack limit here?
Perhaps replace with a loop:
```
for (int i = 0; i < n; i++)
    builder.AppendLine("struct MyStruct{i}<T{i}> {");
for (int i = 0; i < n; i++)
    builder.AppendLine("}");

Reply:
Haven't hit it locally and I went all the way to 20,000. I can fix though. 

---
In reply to: [263978256](https://github.com/dotnet/roslyn/pull/33985#discussion_r263978256) [](ancestors = 263978256)